napari/layers/source.py

```diff
@@ -0,0 +1,23 @@
+from typing import Optional
+
+from pydantic.dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class Source:
+    """An object to store the provenance of a layer.
+
+    Parameters
+    ----------
+    path: str, optional
+        filpath/url associated with layer
+    plugin: str, optional
+        name of plugin that loaded the file (if applicable)
+    method: str, optional
+        fully qualified name of the highest level API method that created the
+        layer (e.g. viewer.open)
+    """
+
+    path: Optional[str] = None
+    plugin: Optional[str] = None
+    method: Optional[str] = None
```

Comment:
Looks like right now the method is just `viewer.open`, `viewer.open_sample`, or one of the Layer init methods themselves `Image.__init__` etc. You'll never get `viewer.add_image` (which is maybe fine). I just want to make sure I understand what's going on.

I don't really have any objections to this, but I also don't really see how/ why someone would use this

Reply:
Use what?  The first challenge we’re trying to solve here is knowing the path/plugin that got the data into the viewer.  If someone uses add_image, there is neither.  If I use the frame stack inspection I described above, I could also put the method in there (so you’d know that add_image was used), but the rest of the source would still be blank.

Reply:
sorry, i meant use the `source.method` string specifically. The `path` and `plugin` strings to me make sense to me, and i guess will be `None` if no path is used, or no `plugin` is used, but I don't quite know what I'd additionally do with `source.method` (as I said, no objection here, just trying to understand).

Reply:
I think @tlambert03 was being forward looking for the stack inspection approach, where `method` contains semi-useful information. I agree with you that it is an awkward thing to add in the earlier commits.

Reply:
`method` removed